pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/ApexUnitTestMethodShouldHaveIsTestAnnotationRule.java

```diff
@@ -0,0 +1,82 @@
+import net.sourceforge.pmd.lang.apex.ast.ASTModifierNode;
+import net.sourceforge.pmd.lang.apex.ast.ASTUserClass;
+import net.sourceforge.pmd.lang.apex.ast.ApexNode;
+import net.sourceforge.pmd.lang.apex.rule.AbstractApexUnitTestRule;
+
+import apex.jorje.semantic.ast.modifier.Annotation;
+import apex.jorje.semantic.ast.modifier.ModifierOrAnnotation;
+import apex.jorje.semantic.symbol.type.AnnotationTypeInfos;
+import apex.jorje.semantic.symbol.type.ModifierOrAnnotationTypeInfo;
+import apex.jorje.semantic.symbol.type.TypeInfoEquivalence;
+import apex.jorje.services.Version;
+
+public class ApexUnitTestMethodShouldHaveIsTestAnnotationRule extends AbstractApexUnitTestRule {
+    private static final String TEST = "test";
+
+    public ApexUnitTestMethodShouldHaveIsTestAnnotationRule() {
+        addRuleChainVisit(ASTUserClass.class);
+    }
+
+    @Override
+    public Object visit(final ASTUserClass node, final Object data) {
+        // test methods should have @isTest annotation.
+        final Version classApiVersion = node.getNode().getDefiningType().getCodeUnitDetails().getVersion();
+
+        if (!isTestMethodOrClass(node) && classApiVersion.isGreaterThan(Version.V174)) {
+            return data;
+        }
+
+        checkForIsTestAnnotation(node, data);
+        return data;
+    }
+
+    private Object checkForIsTestAnnotation(final ApexNode<?> node, final Object data) {
+        final List<ASTMethod> methods = node.findDescendantsOfType(ASTMethod.class);
+        final List<ASTMethod> testMethods = new ArrayList<>();
+        for (ASTMethod method : methods) {
+            final Version classApiVersion = method.getNode().getDefiningType().getCodeUnitDetails().getVersion();
+            if (!isTestMethodOrClass(method) && classApiVersion.isGreaterThan(Version.V174)
+                    && !method.getImage().toLowerCase(Locale.ROOT).contains(TEST)) {
+                continue;
+            }
+            testMethods.add(method);
+        }
+        Map<Integer, ASTMethod> methodLocMap = new HashMap<>();
+        for (ASTMethod testMethod : testMethods) {
+            methodLocMap.put(testMethod.getNode().getLoc().getLine(), testMethod);
+        }
+        List<ASTModifierNode> modifierList = node.findDescendantsOfType(ASTModifierNode.class);
+        final Map<Integer, ModifierOrAnnotation> modifierLocMap = new HashMap<>();
+        for (ASTModifierNode modifier : modifierList) {
+            for (final ModifierOrAnnotationTypeInfo modifierOrAnnotationTypeInfo : modifier.getNode().getModifiers().all()) {
+                ModifierOrAnnotation modifierOrAnnotation = modifier.getNode().getModifiers().get(modifierOrAnnotationTypeInfo);
+                if (modifierOrAnnotation instanceof Annotation && TypeInfoEquivalence
+                        .isEquivalent(modifierOrAnnotationTypeInfo, AnnotationTypeInfos.IS_TEST)) {
+                    modifierLocMap.put(modifierOrAnnotation.getLoc().getLine(), modifierOrAnnotation);
```

Comment:
Since the location of the annoation (the line number) is irrelevant, we should not check this at all.

Consider this code snippet:

```
@isTest
private class ATest {
    @isTest
    @SuppressWarnings
    static void methodATest() {
    }
    static void methodBTest() {
    }
    @isTest static void methodCTest() {
    }
    private void fetchData() {
    }
}
```

Since the annotation is not directly on the line before, we don't see, that `methodATest` is actually annotated.

If you look at the structure of the AST:
![grafik](https://user-images.githubusercontent.com/1573684/53693437-b14a2980-3da0-11e9-9407-0d8de12f3310.png)

It should be enough to just check for the existence of the "IsTest" annoation. 

Now I understand your approach: any method having "test" in its name and that is in a class, that is annotated with "IsTest", should be annotated.

So, first step: collecting the methods:

```java
    private Object checkForIsTestAnnotation(final ApexNode<?> node, final Object data) {
        final List<ASTMethod> methods = node.findDescendantsOfType(ASTMethod.class);
        final List<ASTMethod> testMethods = new ArrayList<>();
        for (ASTMethod method : methods) {
            final Version classApiVersion = method.getNode().getDefiningType().getCodeUnitDetails().getVersion();
            if (!isTestMethodOrClass(method) && classApiVersion.isGreaterThan(Version.V174)
                    && method.getImage().toLowerCase(Locale.ROOT).contains(TEST)) {
                testMethods.add(method);
            }
        }
```

second step: reporting the violations:
```java
        for (ASTMethod method : testMethods) {
            addViolationWithMessage(data, node,
                  "''{0}'' method should have @IsTest annotation.",
                  new Object[] { method.getImage() });
        }
        return data;
    }
```

No need to check for line numbers.

I noticed now, that the rule can easily be written as a XPath expression:
```
//UserClass[ModifierNode/Annotation[@Image = 'IsTest']]
  /Method
    [matches(@Image, 'test', 'i')]
    [not(ModifierNode/Annotation[@Image = 'IsTest'])]
```

See https://pmd.github.io/pmd-6.12.0/pmd_userdocs_extending_writing_xpath_rules.html .
Maybe it's worth to rewrite this rule as a XPath rule? (the test cases stay the same, just the java class ApexUnitTestMethodShouldHaveIsTestAnnotationRule is not necessary and the entry in bestpractices.xml changes)

Reply:
Caveat: the classApiVersion is not checked in XPath, so that might be a reason, why we need to write this for now as a Java rule

Reply:
@triandicAnt why is `classApiVersion` important? I agree with @adangel that this rule could be written as a few line XPath rule opposed to a complex multi hundred line Java rule.

If there is no severe reason for checking the API version please rewrite the rule as lean XPath rule.

Reply:
We check the "classApiVersion" also in the base test class `AbstractApexUnitTestRule`:
https://github.com/pmd/pmd/blob/5d86217871f086f8223da327099d1d67a6e45dab/pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/AbstractApexUnitTestRule.java#L33

I assume, that the code of one class is always the same version. This means, we don't need to do this check again for the methods.

However, if we switch to XPath rule and want to have exactly the same behavior, then need to expose this version in e.g. ASTUserClass via

```java
    public double getVersion() {
        return node.getDefiningType().getCodeUnitDetails().getVersion().getExternal();
    }
```

And then write the XPath as follows:

```
//UserClass[ModifierNode/Annotation[@Image = 'IsTest']][@Version > 23.0]
  /Method
    [matches(@Image, 'test', 'i')]
    [not(ModifierNode/Annotation[@Image = 'IsTest'])]
```

However, maybe we don't need the version check at all? I looked up the mapping between the different version numbers apex uses (I finally found the mapping). The mapping is in the enum `apex.jorje.services.Version`:

```
V140 = external: 6.0 internal: 140.0
V142 = external: 7.0 internal: 142.0
V144 = external: 8.0 internal: 144.0
V146 = external: 9.0 internal: 146.0
V148 = external: 10.0 internal: 148.0
V150 = external: 11.0 internal: 150.0
V150_1 = external: 11.1 internal: 150.1
V152 = external: 12.0 internal: 152.0
V154 = external: 13.0 internal: 154.0
V156 = external: 14.0 internal: 156.0
V158 = external: 15.0 internal: 158.0
V160 = external: 16.0 internal: 160.0
V162 = external: 17.0 internal: 162.0
V164 = external: 18.0 internal: 164.0
V166 = external: 19.0 internal: 166.0
V168 = external: 20.0 internal: 168.0
V170 = external: 21.0 internal: 170.0
V172 = external: 22.0 internal: 172.0
V174 = external: 23.0 internal: 174.0
V176 = external: 24.0 internal: 176.0
V178 = external: 25.0 internal: 178.0
V180 = external: 26.0 internal: 180.0
V182 = external: 27.0 internal: 182.0
V184 = external: 28.0 internal: 184.0
V186 = external: 29.0 internal: 186.0
V188 = external: 30.0 internal: 188.0
V190 = external: 31.0 internal: 190.0
V192 = external: 32.0 internal: 192.0
V194 = external: 33.0 internal: 194.0
V196 = external: 34.0 internal: 196.0
V198 = external: 35.0 internal: 198.0
V200 = external: 36.0 internal: 200.0
V202 = external: 37.0 internal: 202.0
V204 = external: 38.0 internal: 204.0
V206 = external: 39.0 internal: 206.0
V208 = external: 40.0 internal: 208.0
V210 = external: 41.0 internal: 210.0
V212 = external: 42.0 internal: 212.0
V214 = external: 43.0 internal: 214.0
V216 = external: 44.0 internal: 216.0
V218 = external: 45.0 internal: 218.0
V220 = external: 46.0 internal: 220.0
V222 = external: 47.0 internal: 222.0
current: V220
compiler: V210
max: V220
min: V140
post_release: V212
```

We check against the internal version "174", which corresponds to external version "23.0". Now we are at version 45.0.
The check seems only to be relevant is someone still uses version 23.0... which seems to be a apex version 7 years old... (maybe that's the time, when isTest was introduced?).
I'm also not sure, whether the check works as we wanted: If I remove all the isTest annotations from the code, Jorje still parses it as classApiVersion=46.0 ... (which is actually the upcoming Summer 2019 release of apex).
Maybe we are using the Jorje Parser the wrong way?

(Another crazy idea: If we can correctly determine the api version of a apex compilation unit, we could leverage the minimumLanguageVersion/maximumLanguageVersion properties of the rules maybe?)

In short, what we could do in any case:
* Expose the version, so that it is available via XPath
* Write the rule ApexUnitTestMethodShouldHaveIsTestAnnotation as a XPath rule
* Figure out the version uncertainty later

Reply:
Sounds awesome...